From 9e94dcb419ae39efe1e5a392ffd5dcbd28c8484f Mon Sep 17 00:00:00 2001
Message-Id: <9e94dcb419ae39efe1e5a392ffd5dcbd28c8484f.1421390813.git.chang-joon.lee@intel.com>
In-Reply-To: <55165e324aa18c65d2b754de3f9d891b8a8477e9.1421390813.git.chang-joon.lee@intel.com>
References: <55165e324aa18c65d2b754de3f9d891b8a8477e9.1421390813.git.chang-joon.lee@intel.com>
From: Akash Goel <akash.goel@intel.com>
Date: Wed, 29 Oct 2014 15:13:50 +0530
Subject: [PATCH 3/6] REVERTME [VPG]: drm/i915/vlv: Perform modeset based on
 DPMS state during resume

During resume, modeset was being performed independent of DPMS state which
increased resume time as well as it kept display wells ON. With this change
this modeset will be skipped.

Patch is marked as REVERTME, as we intending to replace this VLV specific
solution with a device agnostic implementation from upstream.

v2: Restricting this change only to VLV.

v3: Updated platform check condition to do modeset based on DPMS state. (Jon)

For: GMINL-2847
Change-Id: I079c61d122d37213e636248bf79b4ad56f640a90
Signed-off-by: Akash Goel <akash.goel@intel.com>
Signed-off-by: Sagar Kamble <sagar.a.kamble@intel.com>
---
 drivers/gpu/drm/i915/i915_drv.c |   34 +++++++++++++++++++++++++++++++---
 1 file changed, 31 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.c b/drivers/gpu/drm/i915/i915_drv.c
index a5f4507..c5af16b 100644
--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -676,6 +676,31 @@ void intel_console_resume(struct work_struct *work)
 	console_unlock();
 }
 
+static bool display_is_on(struct drm_device *dev)
+{
+	struct drm_connector *connector;
+	bool display_is_on = false;
+
+	drm_modeset_lock_all(dev);
+	list_for_each_entry(connector, &dev->mode_config.connector_list, head) {
+		if (!connector->encoder || !connector->encoder->crtc)
+			continue;
+		/*
+		 * If Display wasn't turned off, before going to suspend then
+		 * it should be re-enabled now, as we don't expect the DPMS ON
+		 * call to come in that case
+		 */
+		if (connector->dpms != DRM_MODE_DPMS_OFF) {
+			DRM_DEBUG_KMS("Display was on before suspend\n");
+			display_is_on = true;
+			break;
+		}
+	}
+	drm_modeset_unlock_all(dev);
+
+	return display_is_on;
+}
+
 static int i915_drm_thaw_early(struct drm_device *dev)
 {
 	struct drm_i915_private *dev_priv = dev->dev_private;
@@ -738,9 +763,12 @@ static int __i915_drm_thaw(struct drm_device *dev, bool restore_gtt_mappings)
 
 		intel_modeset_init_hw(dev);
 
-		drm_modeset_lock_all(dev);
-		intel_modeset_setup_hw_state(dev, true);
-		drm_modeset_unlock_all(dev);
+		if (!IS_VALLEYVIEW(dev) || IS_CHERRYVIEW(dev)
+						|| display_is_on(dev)) {
+			drm_modeset_lock_all(dev);
+			intel_modeset_setup_hw_state(dev, true);
+			drm_modeset_unlock_all(dev);
+		}
 
 		/*
 		 * ... but also need to make sure that hotplug processing
-- 
1.7.9.5

